### ISSUE TITLE
Limit resources to 63/64 on sub calls and remove `0` as special case for "take all"

### ISSUE THREAD

**Issue by athei** [paritytech/polkadot-sdk#6846]
As of right now when calling a sub contract on `pallet_revive` we allow the sub call to use up all the resources that the caller specifies that are allowed to be used. However, oh Eth this value is capped by 63/64 of the remaining gas. We want to do the same.

Especially, because Solidity will not be able to limit any other resource than ref_time. Limiting of all resources will be enabled by a pre-compile.

**Comment by rockbmb** [comment]
Hello! Can I help with this?

**Comment by athei** [comment]
Yes. It should be a fairly small change to do. Basically, when we create the nested meter for the sub call we need to limit it by 63/64 of the remaining gas:
https://github.com/paritytech/polkadot-sdk/blob/48c6574b1a32893e6a2113622d009cefda0a5f21/substrate/frame/revive/src/gas.rs#L189

Same for the storage meter:
https://github.com/paritytech/polkadot-sdk/blob/48c6574b1a32893e6a2113622d009cefda0a5f21/substrate/frame/revive/src/storage/meter.rs#L289-L299

But please first verify first if it is really 63/64 by checking geth source code (its easy the have one filer per op code).

I also extended the issue (see title) by another thing because I think it is connected: Right now passing `0` as resource value means "take all". This isn't good because we cannot express limiting the resource to `0` as this can make sense in certain cases. Please get rid of this special case (for both meters) and just treat it as "no resource usage allowed". Take all is already accurately represented by just passing a u64::MAX as we are limiting the value anyways to what is available.

**Comment by rockbmb** [comment]
> But please first verify first if it is really 63/64 by checking geth source code (its easy the have one filer per op code).

I was unaware of this mechanism in Ethereum, so I looked it up - https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md

I've searched at https://github.com/ethereum/go-ethereum to confirm the `63/64` value, as you suggested.

In `go-ethereum/core/vm/gas_table.go`, I see
1. [`func gasCall`](https://github.com/ethereum/go-ethereum/blob/88cbfab332c96edfbe99d161d9df6a40721bd786/core/vm/gas_table.go#L373C1-L373C13)
2. [`func gasCallCode`](https://github.com/ethereum/go-ethereum/blob/88cbfab332c96edfbe99d161d9df6a40721bd786/core/vm/gas_table.go#L416C1-L416C17) 
3. [`func gasDelegateCall`](https://github.com/ethereum/go-ethereum/blob/88cbfab332c96edfbe99d161d9df6a40721bd786/core/vm/gas_table.go#L451C2-L451C21)
4. [`func gasStaticCall`](https://github.com/ethereum/go-ethereum/blob/88cbfab332c96edfbe99d161d9df6a40721bd786/core/vm/gas_table.go#L467C6-L467C19)
5. [`func gasSelfdestruct`](https://github.com/ethereum/go-ethereum/blob/88cbfab332c96edfbe99d161d9df6a40721bd786/core/vm/gas_table.go#L483C1-L483C21)

All of the above have a `evm.chainRules.IsEIP150` check, which when enabled, alters the behavior of [`func callGas`](https://github.com/ethereum/go-ethereum/blob/master/core/vm/gas.go#L38C1-L38C13) for functions 1.-4., and for `gasSelfdestruct`, alters the cost of [`SELFDESTRUCT`](https://github.com/ethereum/go-ethereum/blob/master/params/protocol_params.go#L112) from `0` (pre-EIP150) to `5000`.

`func callGas` is similar to `fn nested` you linked, since it determines a function-scoped [`callGasTemp`](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L111C1-L114C20) variable:

```
	// callGasTemp holds the gas available for the current call. This is needed because the
	// available gas is calculated in gasCall* according to the 63/64 rule and later
	// applied in opCall*.
	callGasTemp uint64
```

TL;DR 63/64 is still the number to use.
